Firestore/Swift/Tests/Codable/CodableDocumentTests.swift

```diff
@@ -0,0 +1,403 @@
+          return
+        }
+        if let value = try? values.decode(String.self, forKey: .text) {
+          self = .text(value)
+          return
+        }
+        if let value = try? values.decode(Timestamp.self, forKey: .timestamp) {
+          self = .timestamp(value)
+          return
+        }
+        throw DecodingError.decoding("Decoding error: \(dump(values))")
+      }
+
+      func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        switch self {
+        case let .num(number):
+          try container.encode(number, forKey: .num)
+        case let .text(value):
+          try container.encode(value, forKey: .text)
+        case let .timestamp(stamp):
+          try container.encode(stamp, forKey: .timestamp)
+        }
+      }
+    }
+    struct Model: Codable {
+      let x: Int
+      let e: MyEnum
+    }
+    let model = Model(x: 42, e: MyEnum.num(number: 4))
+    let output = roundTrip(input: model, expected: [:], doTest: false)
+    XCTAssertEqual(model.x, output.x)
+    XCTAssertEqual(model.e, output.e)
+
+    let model2 = Model(x: 43, e: MyEnum.text("abc"))
+    let output2 = roundTrip(input: model2, expected: [:], doTest: false)
+    XCTAssertEqual(model2.x, output2.x)
+    XCTAssertEqual(model2.e, output2.e)
+
+    let model3 = Model(x: 43, e: MyEnum.timestamp(Timestamp(date: Date())))
+    let output3 = roundTrip(input: model3, expected: [:], doTest: false)
+    XCTAssertEqual(model3.x, output3.x)
+    XCTAssertEqual(model3.e, output3.e)
+  }
+
+  func testGeoPoint() {
+    struct Model: Codable {
+      let p: GeoPoint
+    }
+    let model = Model(p: GeoPoint(latitude: 1, longitude: -2))
+    let dict = ["p": GeoPoint(latitude: 1, longitude: -2)]
+    XCTAssertEqual(model.p, roundTrip(input: model, expected: dict).p)
+  }
+
+  func testDate() {
+    struct Model: Codable {
+      let date: Date
+    }
+    let d = Date(timeIntervalSinceReferenceDate: 0)
+    let model = Model(date: d)
+    let dict = ["date": d]
+    XCTAssertEqual(model.date, roundTrip(input: model, expected: dict).date)
+  }
+
+  func testDocumentReference() {
+    struct Model: Codable {
+      let doc: DocumentReference
+    }
+    let d = FSTTestDocRef("abc/xyz")
+    let model = Model(doc: d)
+    let dict = ["doc": d]
+    XCTAssertEqual(model.doc, roundTrip(input: model, expected: dict).doc)
+  }
+
+  func testTimestamp() {
+    struct Model: Codable {
+      let timestamp: Timestamp
+    }
+    let t = Timestamp(date: Date())
+    let model = Model(timestamp: t)
+    let encoded = (try! Firestore.Encoder().encode(model))
+    let model2 = try! Firestore.Decoder().decode(Model.self, from: encoded)
+    XCTAssertEqual(model.timestamp, model2.timestamp)
+  }
+
+  func testBadValue() {
+    struct Model: Codable {
+      let x: Int
+    }
+    let dict = ["x": "abc"]
+    var didThrow = false
+    do {
+      _ = try Firestore.Decoder().decode(Model.self, from: dict)
```

Comment:
I think we should be able to use `XCTAssertThrows`, unless there's an issue with the non discardable return value?